auditing/timescaledb.go

```diff
@@ -0,0 +1,345 @@
+
+	sqlCompOp string
+)
+
+const (
+	equals sqlCompOp = "equals"
+	like   sqlCompOp = "like"
+	phrase sqlCompOp = "phrase"
+)
+
+func NewTimescaleDB(c Config, tc TimescaleDbConfig) (Auditing, error) {
+	if c.Component == "" {
+		component, err := defaultComponent()
+		if err != nil {
+			return nil, err
+		}
+
+		c.Component = component
+	}
+
+	if tc.Port == "" {
+		tc.Port = "5432"
+	}
+	if tc.DB == "" {
+		tc.DB = "postgres"
+	}
+	if tc.User == "" {
+		tc.User = "postgres"
+	}
+	if tc.Retention == "" {
+		tc.Retention = "14 days"
+	}
+	if tc.ChunkInterval == "" {
+		tc.ChunkInterval = "1 days"
+	}
+	if tc.CompressionInterval == "" {
+		tc.CompressionInterval = "7 days"
+	}
+
+	source := fmt.Sprintf("host=%s port=%s user=%s dbname=%s password=%s sslmode=disable", tc.Host, tc.Port, tc.User, tc.DB, tc.Password)
+
+	db, err := sqlx.Connect("postgres", source)
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to datastore: %w", err)
+	}
+
+	a := &timescaleAuditing{
+		component: c.Component,
+		log:       c.Log.WithGroup("auditing"),
+		db:        db,
+		config:    &tc,
+	}
+
+	err = a.initialize()
+	if err != nil {
+		return nil, fmt.Errorf("unable to initialize timescaledb backend: %w", err)
+	}
+
+	a.log.Info("connected to timescaledb backend")
+
+	return a, nil
+}
+
+func (a *timescaleAuditing) initialize() error {
+	type txStatement struct {
+		query string
+		args  []any
+	}
+
+	initialSchema := &migrator.Migration{
+		Name: "Initial database schema",
+		Func: func(tx *sql.Tx) error {
+			for _, stmt := range []txStatement{
+				{
+					query: `CREATE EXTENSION IF NOT EXISTS timescaledb`,
+				},
+				{
+					query: `CREATE EXTENSION IF NOT EXISTS pg_stat_statements`,
+				},
+				{
+					query: `CREATE TABLE IF NOT EXISTS traces (
+						timestamp timestamp NOT NULL,
+						entry jsonb NOT NULL
+					)`,
+				},
+				{
+					query: `ALTER TABLE traces ADD COLUMN ts tsvector GENERATED ALWAYS AS (to_tsvector('simple', entry)) STORED;
+`,
+				},
+				{
+					query: `SELECT create_hypertable('traces', 'timestamp', chunk_time_interval => $1::interval, if_not_exists => TRUE)`,
+					args:  []any{a.config.ChunkInterval},
+				},
+				{
+					query: `ALTER TABLE traces SET (
+						timescaledb.compress,
+						timescaledb.compress_orderby = 'timestamp'
+					)`,
+				},
+				{
+					query: `SELECT add_compression_policy('traces', $1::interval)`,
+					args:  []any{a.config.CompressionInterval},
+				},
+				{
+					query: `CREATE INDEX IF NOT EXISTS traces_gin_idx ON traces USING GIN (entry)`,
+				},
+				{
+					query: `CREATE INDEX IF NOT EXISTS ts_idx ON traces USING GIN (ts)`,
+				},
+				{
+					query: `SELECT add_retention_policy('traces', $1::interval)`,
+					args:  []any{a.config.Retention},
+				},
+			} {
+				if _, err := tx.Exec(stmt.query, stmt.args...); err != nil {
+					return err
+				}
+			}
+
+			return nil
+		},
+	}
+
+	a.db.SetMaxIdleConns(5)
+	a.db.SetConnMaxLifetime(2 * time.Minute)
+	a.db.SetMaxOpenConns(95)
```

Comment:
Hard coded performance tweaks should not be in a library

Reply:
Now configurable from the outside.